include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td

```diff
@@ -272,6 +272,23 @@ def AndOfSelf : Pat <
   (MoveNameHint $old, (AsUIntPrimOp $x)),
   [(KnownWidth $x)]>;
 
+def TypeWidthAdjust32 : NativeCodeCall<
+  "$_builder.getI32IntegerAttr($0.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel() - $1.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel())">;
+
+/// and(x, pad(y, n)) -> pad(and(tail(x), y), n), x is unsigned
+/// And takes an unsigned, so pad must produce an unsigned, which means it's input is unsigned
```

Comment:
And can go from sint to uint apparently?

```firrtl
circuit AndPad:
  module AndPad:
    input x : SInt<3>
    input y : SInt<3>
    output out : UInt

    out <= and(x, pad(y, 4))
```
(which causes verifier failure with this PR)

So I think need to add constraint that x is unsigned (and rest will be per the chain described in comment + existing EqualTypes constraint).